File: app/src/main/java/com/leadinsource/dailyweightlog/SettingsActivity.java
package com.leadinsource.dailyweightlog;

import android.app.Activity;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.MenuItem;

import com.leadinsource.dailyweightlog.app.DWLApplication;

import javax.inject.Inject;

public class SettingsActivity extends AppCompatActivity implements SharedPreferences.OnSharedPreferenceChangeListener {

    private static final String TAG = "Settings Activity";
    private boolean isChanged = false;

    @Inject
    SharedPreferences defaultSharedPreferences;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_settings);

        DWLApplication.app().appComponent().inject(this);
    }

    @Override
    protected void onResume() {
        super.onResume();
        defaultSharedPreferences.registerOnSharedPreferenceChangeListener(this);
        Log.d(TAG, "Reading preferences");
        Log.d(TAG, getResources().getString(R.string.pref_uses_fat_pc_key) + ": " +
                defaultSharedPreferences.getBoolean(getResources().getString(R.string.pref_uses_fat_pc_key),
                        getResources().getBoolean(R.bool.pref_uses_fat_pc_default)));
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {

        switch (item.getItemId()) {
            case android.R.id.home:
                Log.d(TAG, "clicked home button");
                setResult(isChanged);
                break;

        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        isChanged = true;
    }

    @Override
    protected void onPause() {
        super.onPause();
        defaultSharedPreferences.unregisterOnSharedPreferenceChangeListener(this);
    }

    private void setResult(boolean isChanged) {
        Log.d(TAG, "setResult called");
        Intent result = new Intent();
        result.putExtra(MainActivity.PREFERENCES_CHANGED, isChanged);
        setResult(Activity.RESULT_OK, result);
    }

    @Override
    public void finish() {
        setResult(isChanged);
        super.finish();
    }
}

File: app/src/main/java/com/leadinsource/dailyweightlog/SettingsActivity.kt
package com.leadinsource.dailyweightlog

import android.app.Activity
import android.content.Intent
import android.content.SharedPreferences
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import android.util.Log
import android.view.MenuItem

import com.leadinsource.dailyweightlog.app.DWLApplication
import com.leadinsource.dailyweightlog.ui.main.MainActivity

import javax.inject.Inject

private const val TAG = "Settings Activity"

class SettingsActivity : AppCompatActivity(), SharedPreferences.OnSharedPreferenceChangeListener {
    private var isChanged = false


    var defaultSharedPreferences: SharedPreferences? = null
        @Inject set

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_settings)

        DWLApplication.app().appComponent().inject(this)
    }

    override fun onResume() {
        super.onResume()
        defaultSharedPreferences?.registerOnSharedPreferenceChangeListener(this)
        Log.d(TAG, "Reading preferences")
        Log.d(TAG, resources.getString(R.string.pref_uses_fat_pc_key) + ": " +
                defaultSharedPreferences?.getBoolean(resources.getString(R.string.pref_uses_fat_pc_key),
                        resources.getBoolean(R.bool.pref_uses_fat_pc_default)))
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {

        when (item.itemId) {
            android.R.id.home -> {
                Log.d(TAG, "clicked home button")
                setResult(isChanged)
            }
        }

        return super.onOptionsItemSelected(item)
    }

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: String) {
        isChanged = true
    }

    override fun onPause() {
        super.onPause()
        defaultSharedPreferences?.unregisterOnSharedPreferenceChangeListener(this)
    }

    private fun setResult(isChanged: Boolean) {
        Log.d(TAG, "setResult called")
        val result = Intent()
        result.putExtra(MainActivity.PREFERENCES_CHANGED, isChanged)
        setResult(Activity.RESULT_OK, result)
    }

    override fun finish() {
        setResult(isChanged)
        super.finish()
    }
}
